internal/clientconn/cursor/cursor.go

```diff
@@ -27,38 +27,69 @@ import (
 	"sync"
 	"time"
 
+	"go.uber.org/zap"
+
 	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/iterator"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
 	"github.com/FerretDB/FerretDB/internal/util/resource"
 )
 
-// Cursor allows clients to iterate over a result set.
+//go:generate ../../../bin/stringer -linecomment -type Type
+
+// Type represents a cursor type.
+type Type int
+
+const (
+	_ Type = iota
+
+	// Normal represents a normal cursor.
+	Normal
+
+	// Tailable represents a tailable cursor.
+	Tailable
+
+	// TailableAwait represents a tailable and blocking cursor.
+	TailableAwait
+)
+
+// Cursor allows clients to iterate over a result set (or multiple sets for tailable cursors).
 //
-// It implements types.DocumentsIterator interface by wrapping another iterator with documents
+// It implements types.DocumentsIterator interface by wrapping another iterator
 // with additional metadata and registration in the registry.
 //
-// Closing the cursor removes it from the registry.
+// Closing the cursor closes the underlying iterator.
+// For normal cursors, it also removes it from the registry.
+// Tailable cursors are not removed in that case.
 type Cursor struct {
 	// the order of fields is weird to make the struct smaller due to alignment
 
 	created time.Time
-	iter    types.DocumentsIterator
-	r       *Registry
+	iter    types.DocumentsIterator // protected by m
 	*NewParams
-	token     *resource.Token
-	closed    chan struct{}
-	ID        int64
-	closeOnce sync.Once
+	r            *Registry
+	l            *zap.Logger
+	token        *resource.Token
+	removed      chan struct{} // protected by m
+	ID           int64
+	lastRecordID int64 // protected by m
+	m            sync.Mutex
```

Comment:
If they group nicely together you might want to consider this (if you're not worried about data alignment):

```suggestion
	m            sync.Mutex // protects the following fields
	iter         types.DocumentsIterator
	removed      chan struct{}
	lastRecordID int64
```

Reply:
We have many `Cursor` instances, so yeah, here we care about values being smaller